### PR TITLE
Implement CallSmartContract

### PR DESCRIPTION
**Pull request by Ben-Rey**

### CODE REVIEW

File: packages/massa-web3/src/experimental/smartContract.ts [hunk 1]
```diff
@@ -225,12 +239,138 @@ export class SmartContract {
...
+   * @param operationName - The name of the operation where the gas usage is being checked.
+   * @param maxGas - The initial max gas amount proposed for the operation.
+   * @param min - The minimum allowable gas limit for the operation.
+   * @param max - The maximum allowable gas limit for the operation.
+   *
+   * @returns The validated gas amount.
+   *
+   * @throws MaxGasError if the proposed gas amount is either too high or too low.
+   */
+  async getMaxGas(
```
Comment (peterjah) [inline packages/massa-web3/src/experimental/smartContract.ts:314]: i think there is a naming problem. you have to pass MaxGas in getMaxGas

File: packages/massa-web3/src/experimental/smartContract.ts [hunk 2]
```diff
@@ -225,12 +239,138 @@ export class SmartContract {
...
+    operationName: string,
+    maxGas: bigint,
+    min: bigint,
+    max: bigint
+  ): Promise<bigint> {
+    if (!maxGas) return await this.getGasEstimation()
```
Comment (peterjah) [inline packages/massa-web3/src/experimental/smartContract.ts:320]: not sure its working as expected. maxGas should be optionnal

File: packages/massa-web3/src/experimental/smartContract.ts [hunk 3]
```diff
@@ -225,12 +239,138 @@ export class SmartContract {
...
+   *
+   * @returns The validated gas amount.
+   *
+   * @throws MaxGasError if the proposed gas amount is either too high or too low.
+   */
+  async getMaxGas(
+    operationName: string,
+    maxGas: bigint,
+    min: bigint,
+    max: bigint
```
Comment (peterjah) [inline packages/massa-web3/src/experimental/smartContract.ts:318]: i don't undersant what are these values

File: packages/massa-web3/src/experimental/smartContract.ts [hunk 4]
```diff
@@ -225,12 +239,138 @@ export class SmartContract {
...
+   * @param operationName - The name of the operation for which the fee is being checked. This helps identify the context in errors.
+   * @param fee - The fee proposed for the operation. If no fee is provided, the minimal fee will be used.
+   *
+   * @returns The validated or minimal fee.
+   *
+   * @throws MinimalFeeError if the fee provided is less than the minimum required fee.
+   */
+  async getFee(operationName: string, fee: bigint): Promise<bigint> {
+    const minimalFee = await this.client.getMinimalFee()
+    if (!fee) return minimalFee
```
Comment (peterjah) [inline packages/massa-web3/src/experimental/smartContract.ts:342]: fee parameter should be optionnal

File: packages/massa-web3/src/experimental/smartContract.ts (hunk 3 above)
Comment (peterjah) [reply packages/massa-web3/src/experimental/smartContract.ts:318]: if its constants, why using parameters?

File: packages/massa-web3/src/experimental/smartContract.ts [hunk 5]
```diff
@@ -225,12 +239,138 @@ export class SmartContract {
...
+   * Returns the fee for an operation, ensuring it is at least the minimal fee required.
+   *
+   * @param operationName - The name of the operation for which the fee is being checked. This helps identify the context in errors.
+   * @param fee - The fee proposed for the operation. If no fee is provided, the minimal fee will be used.
+   *
+   * @returns The validated or minimal fee.
+   *
+   * @throws MinimalFeeError if the fee provided is less than the minimum required fee.
+   */
+  async getFee(operationName: string, fee: bigint): Promise<bigint> {
```
Comment (peterjah) [inline packages/massa-web3/src/experimental/smartContract.ts:340]: not a fan of passing the name on parameters just for the error message. you should throw a generic error here, and eventually catch it where its needed to add the operation name in the error message

Comment (peterjah) [reply packages/massa-web3/src/experimental/smartContract.ts:340]: We should have one single way of handling fee for every operation. Make the changes for that Transfers; buy roll/ sell roll etc use this implementation

File: packages/massa-web3/src/experimental/smartContract.ts (hunk 3 above)
Comment (Ben-Rey) [reply packages/massa-web3/src/experimental/smartContract.ts:318]: I wanted the function to be used by other operation where the min or max value of maxGas could be different. Like deploy

File: packages/massa-web3/src/experimental/smartContract.ts (hunk 1 above)
Comment (Ben-Rey) [reply packages/massa-web3/src/experimental/smartContract.ts:314]: Changed

File: packages/massa-web3/src/experimental/smartContract.ts [hunk 6]
```diff
@@ -157,7 +163,9 @@ export class SmartContract {
     client: BlockchainClient,
     contractAddress: string
   ): SmartContract {
-    return new SmartContract(client, contractAddress)
+    // Used to validate the address
+    const address = Address.fromString(contractAddress)
+    return new SmartContract(client, address.toString())
```
Comment (peterjah) [inline packages/massa-web3/src/experimental/smartContract.ts:168]: why not take a Address type in contractAddress parameter instead of a string ?

File: packages/massa-web3/src/experimental/smartContract.ts [hunk 7]
```diff
@@ -225,12 +233,92 @@ export class SmartContract {
...
+    account: Account,
+    functionName: string,
+    parameter: Uint8Array,
+    { fee, maxGas, coins = 0n, periodToLive }: CallOptions
+  ): Promise<Operation> {
+    await this.ensureBalance(account, coins)
+
+    fee = fee ?? (await this.client.getMinimalFee())
+
+    if (!maxGas) maxGas = await this.getGasEstimation()
```
Comment (peterjah) [inline packages/massa-web3/src/experimental/smartContract.ts:254]: for code readability i would propose we always use brackets for code blocks..
if its ok for you we can activate the eslint rule for this:
https://archive.eslint.org/docs/2.0.0/rules/curly

File: packages/massa-web3/src/experimental/smartContract.ts [hunk 8]
```diff
@@ -225,12 +233,92 @@ export class SmartContract {
...
+  }
+
+  /**
+   * Ensures that the account has sufficient balance to perform a specified operation.
+   * If the balance is insufficient, this function throws an InsufficientBalanceError.
+   * @param account - The account whose balance is to be checked.
+   * @param coins - The amount of currency required for the operation.
+   * @throws InsufficientBalanceError if the account balance is less than the required amount.
+   */
+  async ensureBalance(account: Account, coins?: bigint): Promise<void> {
```
Comment (peterjah) [inline packages/massa-web3/src/experimental/smartContract.ts:288]: This function is not directly related to smartContract and should be moved somewhere else

File: packages/massa-web3/src/experimental/smartContract.ts [hunk 9]
```diff
@@ -225,12 +233,92 @@ export class SmartContract {
...
+    if (coins) {
```
Comment (peterjah) [inline packages/massa-web3/src/experimental/smartContract.ts:289]: what the point of making "coins" optionnal? the function should just not be called no ??

Comment (Ben-Rey) [reply packages/massa-web3/src/experimental/smartContract.ts:289]: you're right

**Review by peterjah — APPROVED** [review]

File: packages/massa-web3/src/experimental/smartContract.ts (hunk 3 above)
Comment (peterjah) [reply packages/massa-web3/src/experimental/smartContract.ts]: ok i see, lets try to facilitate the dev experience.
Maybe having an optionnal boolean `isDeploy = false` does the trick? ? and its maybe more easy to use

File: packages/massa-web3/src/experimental/basicElements/serializers/bignum.ts [hunk 10]
```diff
@@ -0,0 +1,158 @@
+import { U64_MAX } from './numbers'
```
Comment (gregLibert) [inline packages/massa-web3/src/experimental/basicElements/serializers/bignum.ts:1]: What about:
- creating u64, u128, u256 types as wrapper of bigint?
- adding a from and to generic function that convert everything (maybe at the index level)?

File: packages/massa-web3/src/experimental/basicElements/serializers/strings.ts [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+  if (!arr.length) {
+    return ''
+  }
+
+  if (typeof window === 'undefined') {
+    // eslint-disable-next-line @typescript-eslint/no-var-requires
+    const StringDecoder = require('string_decoder').StringDecoder
+    const decoder = new StringDecoder('utf-8')
+    return decoder.write(Buffer.from(arr))
+  }
```
Comment (gregLibert) [inline packages/massa-web3/src/experimental/basicElements/serializers/strings.ts:32]: if you trying to identify if you're in the browser or in node, you should use the same test as what is done in the crypto module. Ideally, by calling or making this function more usable.

File: packages/massa-web3/src/experimental/errors/base.ts [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+/**
+ * BaseError extends the native Error class to provide additional metadata handling for richer error information.
+ * This class can serve as a base for more specific error types in a wallet or similar system.
+ */
+export class BaseError extends Error {
+  metaMessages: string[]
+  docsPath?: string
+  code: ErrorCodes
+
+  override name = 'WalletProviderError'
```
Comment (gregLibert) [inline packages/massa-web3/src/experimental/errors/base.ts:28]: Why ae we talking about Wallet Prodvier here ?

File: packages/massa-web3/src/experimental/errors/base.ts [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+  constructor(
+    shortMessage: string,
+    args: BaseErrorParameters = { code: ErrorCodes.UnknownError }
+  ) {
+    super()
+
+    const metaMessageStr =
+      args.metaMessages?.map((msg) => `Meta: ${msg}`).join('\n') || ''
+    const docsMessageStr = args.docsPath
+      ? `Docs: ${args.docsPath} for more information.`
```
Comment (gregLibert) [inline packages/massa-web3/src/experimental/errors/base.ts:38]:
```suggestion
      ? `Docs: see ${args.docsPath} for more information.`
```

File: packages/massa-web3/src/experimental/errors/base.ts [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+      details?: never
+    }
+)
+
+export type BaseErrorType = BaseError & { name: 'WalletProviderError' }
+/**
+ * BaseError extends the native Error class to provide additional metadata handling for richer error information.
+ * This class can serve as a base for more specific error types in a wallet or similar system.
+ */
+export class BaseError extends Error {
```
Comment (gregLibert) [inline packages/massa-web3/src/experimental/errors/base.ts:23]: Error should be the prefix. this will make the code more readable when listing multiple errors

File: packages/massa-web3/src/experimental/errors/base.ts [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+      cause?: never
+      details?: string
+    }
+  | {
+      cause: BaseError | Error
+      details?: never
+    }
+)
+
+export type BaseErrorType = BaseError & { name: 'WalletProviderError' }
```
Comment (gregLibert) [inline packages/massa-web3/src/experimental/errors/base.ts:18]: Should be `ErrorTypeBase` to make the code easier to read.

File: packages/massa-web3/src/experimental/errors/base.ts [hunk 16]
```diff
@@ -0,0 +1,57 @@
+import { ErrorCodes } from './utils/codes'
+
+export type BaseErrorParameters = {
```
Comment (gregLibert) [inline packages/massa-web3/src/experimental/errors/base.ts:3]: Is this long name mandatory?

File: packages/massa-web3/src/experimental/errors/insufficientBalance.ts [hunk 17]
```diff
@@ -0,0 +1,42 @@
+import { BaseError } from './base'
+import { ErrorCodes } from './utils/codes'
+
+type InsufficientBalanceErrorParameters = {
+  userBalance: bigint
+  neededBalance: bigint
+  cause?: Error
+}
+
+export type InsufficientBalanceErrorType = InsufficientBalanceError & {
```
Comment (gregLibert) [inline packages/massa-web3/src/experimental/errors/insufficientBalance.ts:10]: Why do we need to export it?

File: packages/massa-web3/src/experimental/smartContract.ts (hunk 6 above)
Comment (gregLibert) [reply packages/massa-web3/src/experimental/smartContract.ts:168]: I agree with Pierre on this. As a rule of thumb, you should take an address. What use case do you have in mind to justify the string parameter?

File: packages/massa-web3/src/experimental/smartContract.ts [hunk 18]
```diff
@@ -225,12 +233,82 @@ export class SmartContract {
...
+   * @returns A promise that resolves to an Operation object representing the transaction.
+   */
   async call(
-    _method: string, // eslint-disable-line @typescript-eslint/no-unused-vars
-    _args: Uint8Array, // eslint-disable-line @typescript-eslint/no-unused-vars
-    _opts: CallOptions // eslint-disable-line @typescript-eslint/no-unused-vars
-  ): Promise<Uint8Array> {
-    throw new Error('Not implemented')
+    account: Account,
+    functionName: string,
```
Comment (gregLibert) [inline packages/massa-web3/src/experimental/smartContract.ts:246]: `method` or `func` should be used if funtion can be. The name doesn't make things clearer.

File: packages/massa-web3/src/experimental/smartContract.ts [hunk 19]
```diff
@@ -225,12 +233,82 @@ export class SmartContract {
...
+    parameter: Uint8Array,
+    { fee, maxGas, coins = 0n, periodToLive }: CallOptions
```
Comment (gregLibert) [inline packages/massa-web3/src/experimental/smartContract.ts:248]: This complefify the code. you should replace with a `opts` name and handle the coins=0 case in the body.

File: packages/massa-web3/src/experimental/smartContract.ts [hunk 20]
```diff
@@ -225,12 +233,82 @@ export class SmartContract {
...
+   *
+   * @returns The calculated expiration period for the operation.
+   */
+  async getExpirePeriod(periodToLive?: number): Promise<number> {
+    const currentPeriod = await this.client.fetchPeriod()
+    // TODO - What should be the default value ?
+    return !periodToLive
+      ? currentPeriod + MIN_PERIODS_TO_LIVE
+      : calculateExpirePeriod(currentPeriod, periodToLive)
+  }
```
Comment (gregLibert) [inline packages/massa-web3/src/experimental/smartContract.ts:303]: The code can be simplified and `MIN_PERIODS_TO_LIVE` is too short to be used as default value.

```suggestion
 const DEFAULT_PERIODS_TO_LIVE = 9
 ...
  async getExpirePeriod(periodToLive?: number = DEFAULT_PERIODS_TO_LIVE): Promise<number> {
    const currentPeriod = await this.client.fetchPeriod()
    return calculateExpirePeriod(currentPeriod, periodToLive)
  }
```

**Review by gregLibert — CHANGES_REQUESTED** [review]

File: packages/massa-web3/src/experimental/basicElements/serializers/arrays.ts [hunk 21]
```diff
@@ -0,0 +1,244 @@
...
+      return 4
+    case ArrayTypes.F64:
+    case ArrayTypes.I64:
+    case ArrayTypes.U64:
+      return 8
+    case ArrayTypes.I128:
+    case ArrayTypes.U128:
+      return 16
+    case ArrayTypes.U256:
+      return 32
```
[the remark this answers is not in the document]
Comment (Ben-Rey) [reply packages/massa-web3/src/experimental/basicElements/serializers/arrays.ts:47]: It's an assemblyscript feature

File: packages/massa-web3/src/experimental/errors/base.ts (hunk 12 above)
Comment (Ben-Rey) [reply packages/massa-web3/src/experimental/errors/base.ts:28]: wrong copy paste

File: packages/massa-web3/src/experimental/errors/base.ts (hunk 16 above)
Comment (Ben-Rey) [reply packages/massa-web3/src/experimental/errors/base.ts:3]: This one is shorter. It's less descriptive but should be enough as it's only used here

File: packages/massa-web3/src/experimental/errors/insufficientBalance.ts (hunk 17 above)
Comment (Ben-Rey) [reply packages/massa-web3/src/experimental/errors/insufficientBalance.ts:10]: It's lighter if we only need the type which will mostly be the case for dev

File: packages/massa-web3/src/experimental/basicElements/serializers/strings.ts (hunk 11 above)
Comment (Ben-Rey) [reply packages/massa-web3/src/experimental/basicElements/serializers/strings.ts:32]: like this ?

File: packages/massa-web3/src/experimental/basicElements/serializers/arrays.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (gregLibert) [reply packages/massa-web3/src/experimental/basicElements/serializers/arrays.ts:47]: :facepalm:

File: packages/massa-web3/src/experimental/basicElements/serializers/strings.ts [hunk 22]
```diff
@@ -17,25 +19,27 @@ export function strToBytes(str: string): Uint8Array {
...
-  if (typeof window === 'undefined') {
+  if (isNode()) {
     // eslint-disable-next-line @typescript-eslint/no-var-requires
-    const StringDecoder = require('string_decoder').StringDecoder
+    const { StringDecoder } = require('string_decoder')
     const decoder = new StringDecoder('utf-8')
+
     return decoder.write(Buffer.from(arr))
-  }
+  } else {
```
Comment (gregLibert) [inline packages/massa-web3/src/experimental/basicElements/serializers/strings.ts:35]: The else is useless here.

**Review by gregLibert — CHANGES_REQUESTED** [review]

**Review by gregLibert — APPROVED** [review]
